crates/flycheck/src/project_json.rs

```diff
@@ -0,0 +1,155 @@
+//! A `cargo-metadata`-equivalent for non-Cargo build systems.
+use std::{io, process::Command};
+
+use crossbeam_channel::Sender;
+use paths::{AbsPathBuf, Utf8Path, Utf8PathBuf};
+use project_model::ProjectJsonData;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+use crate::command::{CommandHandle, ParseFromLine};
+
+/// A command wrapper for getting a `rust-project.json`.
+///
+/// This is analogous to `cargo-metadata`, but for non-Cargo build systems.
+pub struct Discover {
+    command: Vec<String>,
+    sender: Sender<DiscoverProjectMessage>,
+}
+
+#[derive(PartialEq, Clone, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub enum Arguments {
+    Path(
+        #[serde(serialize_with = "serialize_abs_pathbuf")]
+        #[serde(deserialize_with = "deserialize_abs_pathbuf")]
```

Comment:
> Don't actually need a Deserialize implementation for this type that only exists to be serialized.

Whoops, yup. Will remove.

> Also, since this it copy pasted below, maybe you can just implement the serde traits on AbsPathBuf itself?

I don't want to that because taking a dependency on serde in the crate where `AbsPathBuf` is defined (`paths`) slows down compilation because `serde_derive` now needs to be compiled *way* sooner.